core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePermissionProvider.java

```diff
@@ -44,15 +45,15 @@ private boolean permitCmAndStackUpgradeByBuildNumber(Image currentImage, Image i
         return componentBuildNumberComparator.compare(currentImage, image, buildNumberKey);
     }
 
-    boolean permitCmAndStackUpgradeByComponentVersion(String currentVersion, String newVersion) {
-        return permitByComponentVersion(currentVersion, newVersion) && permitByUpgradeMatrix(currentVersion, newVersion);
+    boolean permitCmAndStackUpgradeByComponentVersion(String currentVersion, String newVersion, boolean checkUpgradeMatrix) {
+        return permitByComponentVersion(currentVersion, newVersion) && permitByUpgradeMatrix(currentVersion, newVersion, checkUpgradeMatrix);
     }
 
     private boolean permitByComponentVersion(String currentVersion, String newVersion) {
         return componentVersionComparator.permitCmAndStackUpgradeByComponentVersion(currentVersion, newVersion);
     }
 
-    private boolean permitByUpgradeMatrix(String currentVersion, String newVersion) {
-        return upgradeMatrixService.permitByUpgradeMatrix(currentVersion, newVersion);
+    private boolean permitByUpgradeMatrix(String currentVersion, String newVersion, boolean checkUpgradeMatrix) {
+        return !checkUpgradeMatrix || upgradeMatrixService.permitByUpgradeMatrix(currentVersion, newVersion);
```

Comment:
!checkUpgradeMatrix is not part of the permit logic hence I would extract it to the containing method: 

 permitByComponentVersion(currentVersion, newVersion) && (!checkUpgradeMatrix || permitByUpgradeMatrix(currentVersion, newVersion, checkUpgradeMatrix) );